src/mongo/Config.class.php

```diff
@@ -1759,12 +1760,47 @@ protected function getConnectionForDataSource($dataSource)
         }
         if(!isset($this->connections[$dataSource]))
         {
-            $this->connections[$dataSource] = new \MongoClient($ds['connection'], $connectionOptions);
+            $retries = 1;
+            $exception = null;
+            $connected = false;
+
+            do {
+                try {
+                    $this->connections[$dataSource] = $this->getMongoClient($ds['connection'], $connectionOptions);
+                    // Check if we are connected
+                    $connections = $this->connections[$dataSource]->getConnections();
+                    if (empty($connections) === false) {
+                        $connected = true;
+                    }
+                } catch (\MongoConnectionException $e) {
+                    self::getLogger()->warn("MongoConnectionException: " . $e->getMessage());
+                    sleep(1);
+                    $retries++;
+                    $exception = $e;
+                }
+
+            } while ($retries <= self::CONNECTION_RETRIES && $connected == false);
+
+            if (!isset($this->connections[$dataSource])) {
+                self::getLogger()->warn("MongoConnectionException after " . $retries . " attempts (MAX:".self::CONNECTION_RETRIES."): " . $e->getMessage());
+                throw new \MongoConnectionException($exception);
+            }
         }
         return $this->connections[$dataSource];
     }
 
     /**
+     * Create a Mongo Client - used for mocking
+     * @param string $connection
+     * @param array $connectionOptions
+     * @return \MongoClient
+     */
+    protected function getMongoClient($connection, $connectionOptions)
+    {
+        return new \MongoClient($connection, $connectionOptions);
```

Comment:
do you want a new mongo client every time? or should this be a lazy load? what (if any ) is the overhead of creating a new client each time?